src/System.Memory/src/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,305 @@
+
+            while (current != null)
+            {
+                // We need to loop up until the end to make sure start and end are connected
+                // if end is not trusted
+                if (!foundResult)
+                {
+                    Memory<byte> memory = current.Memory;
+                    int currentEnd = current == end ? endPosition : memory.Length;
+
+                    memory = memory.Slice(0, currentEnd - currentIndex);
+                    // We would prefer to put position in the beginning of next segment
+                    // then past the end of previous one, but only if we are not leaving current buffer
+                    if (memory.Length > count ||
+                       (memory.Length == count && current == end))
+                    {
+                        result = new SequencePosition(current, currentIndex + (int)count);
+                        foundResult = true;
+                        if (!checkEndReachable)
+                        {
+                            break;
+                        }
+                    }
+
+                    count -= memory.Length;
+                }
+
+                if (current.Next == null && current != end)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                }
+
+                current = current.Next;
+                currentIndex = 0;
+            }
+
+            if (!foundResult)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+            }
+
+            return result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private long GetLength(SequencePosition start, SequencePosition end)
+        {
+            int startIndex = start.Index;
+            int endIndex = end.Index;
+            BufferType type = GetBufferType();
+
+            startIndex = GetIndex(startIndex);
+            endIndex = GetIndex(endIndex);
+
+            switch (type)
+            {
+                case BufferType.MemoryList:
+                    return GetLength((IMemoryList<T>)start.Segment, startIndex, (IMemoryList<T>)end.Segment, endIndex);
+
+                case BufferType.OwnedMemory:
+                case BufferType.Array:
+                    return endIndex - startIndex;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static long GetLength(
+            IMemoryList<T> start,
+            int startIndex,
+            IMemoryList<T> endSegment,
+            int endIndex)
+        {
+            if (start == endSegment)
+            {
+                return endIndex - startIndex;
+            }
+
+            return (endSegment.RunningIndex - start.Next.RunningIndex)
```

Comment:
Can you please add a comment here explaining this calculation and that we know it won't overflow?

Reply:
Why would it overflow, it not addition.

Reply:
There is addition:
```C#
return (endSegment.RunningIndex - start.Next.RunningIndex) + (start.Memory.Length - startIndex) + endIndex;
```